DI-2341 Fix handling of mixed runs to process assays with hold=false first.

RSWilson1 opened the pull request:
PR for fixing https://github.com/eastgenomics/eggd_conductor/issues/172 and https://github.com/eastgenomics/eggd_conductor/issues/169.

Changed the processing order of the assay handlers by sorting `hold=false` first.

Successful job: https://platform.dnanexus.com/panx/projects/J2Z7KF043bx7Kb25J99zzvPy/monitor/job/J2gV2v043bxBqP9K8pK6Z344

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)
<!-- Reviewable:end -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added helpers to detect "hold" flags and to search exact keys in nested configurations; held items are processed last.

- Bug Fixes
  - Ticket, job and platform links now use HTTPS and canonical job IDs/paths for reliable linking.

- Tests
  - New tests covering hold detection and exact-key matching across nested structures.

- Documentation
  - Example URLs in README updated to use HTTPS.

- Chores
  - Application version bumped to 2.3.0.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RSWilson1 committed:
Sort handlers to have hold=true last.

RSWilson1 committed:
Fixed url for demultiplexing by removing "job-" and added https

RSWilson1 committed:
Added sorting of assay handlers by if the config contains hold=true, this processes all hold=false configs first.
Changed eggd_conductor link to https to match DNAnexus.

RSWilson1 committed:
Increment version

Yu-jinKim reviewed (changes requested):
> @Yu-jinKim reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: all files reviewed, 7 unresolved discussions (waiting on @RSWilson1)

___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 1056 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg9OUL-HTwg1dT56CN:-OYg9OUL-HTwg1dT56CO:b-q87bs4) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L1056)):*
> ```Python
> 
> 
> def is_held(config):
> ```

can you add typing ? i.e. config: str and -> bool
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 1073 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYgAgEJ2jAz-2Pv9A89:-OYgAgEK2-jPA7e2IbNt:b-wffntu) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L1073)):*
> ```Python
>     # Check if any of those keys are set to True or "true" in found
>     for value in found:
>         if value is True or (isinstance(value, str) and value.lower() == "true"):
> ```

is that because if you specify an assay config, it gets parsed by json.load and if you don't it's a dnanexus text file ?
___
*[`resources/home/dnanexus/run_workflows/run_workflows.py` line 556 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg8Zdo6Fx8ljy9rp-I:-OYg8Zdo6Fx8ljy9rp-J:bq4gp1g) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/run_workflows.py#L556)):*
> ```Python
>     assay_handlers = sorted(
>         assay_handlers, key=lambda h: manage_dict.is_held(h.config)
>         )
> ```

minor: can that closing parentheses be deindented ?
___
*[`resources/home/dnanexus/run_workflows/run_workflows.py` line 557 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg8s7OAld3weRTR6U6:-OYg8s7OAld3weRTR6U7:bata78e) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/run_workflows.py#L557)):*
> ```Python
>         assay_handlers, key=lambda h: manage_dict.is_held(h.config)
>         )
>     print("\nAssay handler order (those with hold=True are last):")
> ```

do you plan on keeping that print? if so can you use prettier_print ?


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
added unit tests for is_held().
added new function for search_exact_key() to prevent threshold or other *hold being recognised in is_held().

RSWilson1 committed:
rm print as already can see order in other print

RSWilson1 committed:
Added tests for search_exact_key()

RSWilson1 commented on resources/home/dnanexus/run_workflows/utils/manage_dict.py:
```diff
@@ -1051,3 +1051,25 @@ def populate_tso500_reports_workflow(
+    for value in found:
+        if value is True or (isinstance(value, str) and value.lower() == "true"):
+            return True
+    return False
```
Fixed by adding search_exact_key() for is_held()
<!-- Reviewable comment -OYkBuHS4SNArMYqKXWB:b3giizo -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: 2 of 5 files reviewed, 7 unresolved discussions (waiting on @Yu-jinKim)

___
*[`resources/home/dnanexus/run_workflows/run_workflows.py` line 556 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg8Zdo6Fx8ljy9rp-I:-OYkBS8kA_9n_-L5qmjS:b-n9tkhd) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/run_workflows.py#L556)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

minor: can that closing parentheses be deindented ?
</blockquote></details>

changed
___
*[`resources/home/dnanexus/run_workflows/run_workflows.py` line 557 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg8s7OAld3weRTR6U6:-OYkBOvyBsJ2PnD7DB_F:bb7nx99) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/run_workflows.py#L557)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

do you plan on keeping that print? if so can you use prettier_print ?
</blockquote></details>

rm'd
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 1056 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg9OUL-HTwg1dT56CN:-OYkIB7oDmP9ToHDcwTx:b-7tfmjp) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L1056)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

can you add typing ? i.e. config: str and -> bool
</blockquote></details>

added
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 1073 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYgAgEJ2jAz-2Pv9A89:-OYkIJrsBO76Egsvix9s:bhrj7gy) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L1073)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

is that because if you specify an assay config, it gets parsed by json.load and if you don't it's a dnanexus text file ?
</blockquote></details>

I'm not sure I get what you mean? This is handling if the input is a bool or a str of true/false.


<!-- Sent from Reviewable.io -->

Yu-jinKim reviewed (changes requested):
> @Yu-jinKim reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: all files reviewed, 8 unresolved discussions (waiting on @RSWilson1)

___
*[`resources/home/dnanexus/run_workflows/run_workflows.py` line 556 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg8Zdo6Fx8ljy9rp-I:-OYkTQMp7k4VuEQ-HbJ9:bk3ve0v) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/run_workflows.py#L556)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

changed
</blockquote></details>

i still see it indented
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 1073 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYgAgEJ2jAz-2Pv9A89:-OYkWCEQ65dAUvW3uIoT:beslofz) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L1073)):*
<details><summary><i>Previously, RSWilson1 (Robert Wilson) wrote…</i></summary><blockquote>

I'm not sure I get what you mean? This is handling if the input is a bool or a str of true/false.
</blockquote></details>

i am wondering why we need to handle both booleans and strings as i would expect only booleans to be set in the assay configs
___
*[`resources/home/dnanexus/run_workflows/tests/test_manage_dict.py` line 1476 at r2](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYkVD3nEbeifNcPcKxM:-OYkVD3nEbeifNcPcKxN:b6rjp4h) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/2891d78ea36292632e38073e59bb08ceb898d9c3/resources/home/dnanexus/run_workflows/tests/test_manage_dict.py#L1476)):*
> ```Python
> 
>     def test_very_nested_hold_true_str(self):
>         """Test when hold key is very nested and True bool"""
> ```

wrong docstring
___
*[`resources/home/dnanexus/run_workflows/tests/test_manage_dict.py` line 1477 at r2](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYkULJh2yEC9oAUhBnF:-OYkULJh2yEC9oAUhBnG:beouplr) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/2891d78ea36292632e38073e59bb08ceb898d9c3/resources/home/dnanexus/run_workflows/tests/test_manage_dict.py#L1477)):*
> ```Python
>     def test_very_nested_hold_true_str(self):
>         """Test when hold key is very nested and True bool"""
>         config = {"foo": {"bar": {"hold": "true"}}}
> ```

a more appropriate name would be `test_nested_hold_true_str` since it is the same level of nestedness as the `test_nested_hold_true` function
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 68 at r2](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYkVxrREvucri9mXkjs:-OYkVxrSBdYjsYAm18e3:b-dnhg59) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/2891d78ea36292632e38073e59bb08ceb898d9c3/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L68)):*
> ```Python
> 
> 
> def search_exact_key(identifier, input_dict) -> list:
> ```

can you add typing for the arguments?


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
change to single line

RSWilson1 committed:
add typing for inputs

RSWilson1 committed:
Rm'd handling for true/false strings
Changed test names and rm'd tests for strings

RSWilson1 committed:
Fixed typo in docstring
Added test with tso500 config

RSWilson1 committed:
Updated docstring to only handling bools

RSWilson1 committed:
Made code more concise.

RSWilson1 committed:
Added http to https

RSWilson1 committed:
Changed to handle unhashable values

RSWilson1 committed:
Standardised file path handling for test

RSWilson1 committed:
Updates links to match https

RSWilson1 commented on resources/home/dnanexus/run_workflows/utils/manage_dict.py:
```diff
@@ -1051,3 +1077,25 @@ def populate_tso500_reports_workflow(
+    for value in found:
+        if value is True:
+            return True
+    return False
```
Fixed docstring and used logic for return statement.
<!-- Reviewable comment -OYl8LSz6a8YP8p12K03:bkkljr8 -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: 2 of 6 files reviewed, 9 unresolved discussions (waiting on @Yu-jinKim)

___
*[`resources/home/dnanexus/run_workflows/run_workflows.py` line 556 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYg8Zdo6Fx8ljy9rp-I:-OYkdQnO40jMtZjDE5_p:b-tby1ny) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/run_workflows.py#L556)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

i still see it indented
</blockquote></details>

awh yes it is, changed to a single line as formatter recommended.
___
*[`resources/home/dnanexus/run_workflows/tests/test_manage_dict.py` line 1476 at r2](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYkVD3nEbeifNcPcKxM:-OYl7erc4oq92WDP8pTx:b-a2ujfl) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/2891d78ea36292632e38073e59bb08ceb898d9c3/resources/home/dnanexus/run_workflows/tests/test_manage_dict.py#L1476)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

wrong docstring
</blockquote></details>

Changed.
___
*[`resources/home/dnanexus/run_workflows/tests/test_manage_dict.py` line 1477 at r2](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYkULJh2yEC9oAUhBnF:-OYl7fnC02bStlVorSp_:b-a2ujfl) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/2891d78ea36292632e38073e59bb08ceb898d9c3/resources/home/dnanexus/run_workflows/tests/test_manage_dict.py#L1477)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

a more appropriate name would be `test_nested_hold_true_str` since it is the same level of nestedness as the `test_nested_hold_true` function
</blockquote></details>

Changed.
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 1073 at r1](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYgAgEJ2jAz-2Pv9A89:-OYl7rHABrAySUkp6BTF:bu13yu2) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/066cda8e20493a45b8b9ab64a04f67d343412f6b/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L1073)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

i am wondering why we need to handle both booleans and strings as i would expect only booleans to be set in the assay configs
</blockquote></details>

Ahhh I see, yes I guess we shouldn't so I have rm'd the handling for that.
___
*[`resources/home/dnanexus/run_workflows/utils/manage_dict.py` line 68 at r2](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYkVxrREvucri9mXkjs:-OYkdG4O7h33Cr2xELtE:b-7tfmjp) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/2891d78ea36292632e38073e59bb08ceb898d9c3/resources/home/dnanexus/run_workflows/utils/manage_dict.py#L68)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

can you add typing for the arguments?
</blockquote></details>

added


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
Reverted to original code as clearer and don't need handling of non-hashable types.
Added additional tests

RSWilson1 committed:
Renamed tests and made data more relevant

RSWilson1 committed:
Updated docstring to match test

Yu-jinKim reviewed (changes requested):
> @Yu-jinKim reviewed 1 of 3 files at r4, 1 of 2 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: all files reviewed, 6 unresolved discussions (waiting on @RSWilson1)

___
*[`resources/home/dnanexus/run_workflows/tests/test_manage_dict.py` line 1433 at r7](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYpRHVDEqABRLrHkiZv:-OYpRHVDEqABRLrHkiZw:b-7k93k5) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/f6e5078a37d431b72f997c5f1602357fe8ebbacc/resources/home/dnanexus/run_workflows/tests/test_manage_dict.py#L1433)):*
> ```Python
>     """
> 
>     def test_no_hold_key_present(self):
> ```

is it too late to ask for a parametrize decorator ?
it would ease comprehension and you could add a lot of test cases i.e. lists, dicts
sorry i should have thought of that in the first pass


<!-- Sent from Reviewable.io -->

RSWilson1 committed:
Changed to parameterized tests and added some additional test cases.

RSWilson1 committed:
rm debug print

RSWilson1 commented on resources/home/dnanexus/run_workflows/run_workflows.py:
```diff
@@ -398,7 +398,7 @@ def main():
                             "eggd_conductor: "
                         ),
-                        url=f"http://{os.environ.get('conductor_job_url')}",
+                        url=f"https://{os.environ.get('conductor_job_url')}",
                         ticket=ticket["id"],
                     )
```
This isn't relevant to my changes.
<!-- Reviewable comment -OYpwHFXDCOaxdhSEkfB:bredm29 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on resources/home/dnanexus/run_workflows/tests/test_manage_dict.py:
```diff
@@ -1176,18 +1125,16 @@ def test_array_input_with_single_dict_given(self):
         input_type = type(
             output["stage-G9Z2B8841bQY907z1ygq7K9x.somalier_extract_file"]
         )
 
-        assert (
-            input_type == list
-        ), "array:file input not converted to a list as expected"
+        assert input_type == list, (
+            "array:file input not converted to a list as expected"
+        )
```
This isn't relevant to my changes.
<!-- Reviewable comment -OYpwNOVDHqHKdgIWNhV:bredm29 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on resources/home/dnanexus/run_workflows/tests/test_manage_dict.py:
```diff
@@ -1481,3 +1422,153 @@ def test_missing_metrics_output_raises_assertion_error(self):
+        """Keys containing substring do not match"""
+        config = {"app": "app-id", "hold": {"value": True}}
+        assert search_exact_key("hold", config) == []
+
```
not needed
<!-- Reviewable comment -OYpuZYOAMjnu1YedQ2c:bpp6stj -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on resources/home/dnanexus/run_workflows/tests/test_manage_dict.py:
```diff
@@ -1481,3 +1422,255 @@ def test_missing_metrics_output_raises_assertion_error(self):
+            # Keys containing substring do not match
+            ({"threshold": 0.5, "foo": {"holdings": True}}, "hold", []),
+            # Keys containing list of ints do not match
+            ({"app": "app-id", "details": {"hold": [1, 2, 3]}}, "hold", []),
+            # Keys containing list of bools do not match
+            ({"app": "app-id", "details": {"hold": [True, False, False]}}, "hold", []),
+            # Keys of hold with value as dict do not match
+            ({"app": "app-id", "details": {"hold": {"key": True}}}, "hold", []),
+            # List of dicts with matching key (ints)
+            ({"foo": [{"hold": 1}, {"hold": 2}]}, "hold", [1, 2]),
+            # List of dicts with matching key (bools)
+            ({"foo": [{"hold": True}, {"hold": False}]}, "hold", [True, False]),
```
These changes aren't needed
<!-- Reviewable comment -OYpuMMpDYLXrDvqn5qf:b-e3qjv5 -->
<!-- Sent from Reviewable.io -->

RSWilson1 commented on resources/home/dnanexus/run_workflows/utils/demultiplexing.py:
```diff
@@ -317,11 +317,12 @@ def demultiplex(
         # dx job error raised (i.e. failed, timed out, terminated)
+        job_id = job.id.replace("job-", "")
         job_url = (
-            f"platform.dnanexus.com/projects/"
+            f"https://platform.dnanexus.com/projects/"
             f"{demultiplex_project.replace('project-', '')}"
             "/monitor/job/"
-            f"{job.id}"
+            f"{job_id}"
         )
```
I don't think this is needed.
<!-- Reviewable comment -OYpwDa_2U_ID3g_LOLp:b-a6bqt -->
<!-- Sent from Reviewable.io -->

RSWilson1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: 4 of 6 files reviewed, 6 unresolved discussions (waiting on @Yu-jinKim)

___
*[`resources/home/dnanexus/run_workflows/tests/test_manage_dict.py` line 1433 at r7](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175#-OYpRHVDEqABRLrHkiZv:-OYpw50O5T8qxzz_2NCw:b-f98cj8) ([raw file](https://github.com/eastgenomics/eggd_conductor/blob/f6e5078a37d431b72f997c5f1602357fe8ebbacc/resources/home/dnanexus/run_workflows/tests/test_manage_dict.py#L1433)):*
<details><summary><i>Previously, Yu-jinKim wrote…</i></summary><blockquote>

is it too late to ask for a parametrize decorator ?
it would ease comprehension and you could add a lot of test cases i.e. lists, dicts
sorry i should have thought of that in the first pass
</blockquote></details>

Nope, great suggestion :thumbsup: Added.


<!-- Sent from Reviewable.io -->

Yu-jinKim reviewed (approved):
> @Yu-jinKim reviewed 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: all files reviewed, 5 unresolved discussions (waiting on @RSWilson1)



<!-- Sent from Reviewable.io -->

Yu-jinKim reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/eggd_conductor/175)* status: all files reviewed, 5 unresolved discussions (waiting on @RSWilson1)



<!-- Sent from Reviewable.io -->
